cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -1194,6 +1198,197 @@ void AddSplit(FunctionRegistry* registry) {
+          std::search(i, end, options.pattern.begin(), options.pattern.end());
+      if (pos == end) {
+        RETURN_NOT_OK(builder->Append(reinterpret_cast<const uint8_t*>(i),
+                                      static_cast<offset_type>(end - i)));
+        i = end;
+      } else {
+        // the string before the pattern
+        RETURN_NOT_OK(builder->Append(reinterpret_cast<const uint8_t*>(i),
+                                      static_cast<offset_type>(pos - i)));
+        // the replacement
+        RETURN_NOT_OK(
+            builder->Append(reinterpret_cast<const uint8_t*>(options.replacement.data()),
+                            options.replacement.length()));
+        // skip pattern
+        i = pos + options.pattern.length();
+        max_replacements--;
+      }
+    }
+    // if we exited early due to max_replacements, add the trailing part
+    RETURN_NOT_OK(builder->Append(reinterpret_cast<const uint8_t*>(i),
+                                  static_cast<offset_type>(end - i)));
+    return Status::OK();
+  }
+};
+
+const FunctionDoc replace_substring_doc(
+    "Replace non-overlapping substrings that match pattern by replacement",
+    ("For each string in `strings`, replace non-overlapping substrings that match\n"
+     "`pattern` by `replacement`. If `max_replacements != -1`, it determines the\n"
+     "maximum amount of replacements made, counting from the left. Null values emit\n"
+     "null."),
+    {"strings"}, "ReplaceSubstringOptions");
+
+#ifdef ARROW_WITH_RE2
+template <typename Type>
+struct ReplaceSubStringRE2 : ReplaceSubStringBase<Type, ReplaceSubStringRE2<Type>> {
+  using Base = ReplaceSubStringBase<Type, ReplaceSubStringRE2<Type>>;
+  using ValueDataBuilder = typename Base::ValueDataBuilder;
+  using offset_type = typename Base::offset_type;
+
+  ReplaceSubstringOptions options;
+  RE2 regex_find;
+  RE2 regex_replacement;
+  explicit ReplaceSubStringRE2(KernelContext* ctx, ReplaceSubstringOptions options)
+      : options(options),
+        regex_find("(" + options.pattern + ")"),
+        regex_replacement(options.pattern) {
+    // Using RE2::FindAndConsume we can only find the pattern if it is a group, therefore
+    // we have 2 regex, one with () around it, one without.
+    if (!(regex_find.ok() && regex_replacement.ok())) {
+      ctx->SetStatus(Status::Invalid("Regular expression error"));
+      return;
+    }
+  }
+  Status ReplaceString(util::string_view s, ValueDataBuilder* builder) {
+    re2::StringPiece replacement(options.replacement);
+    if (options.max_replacements == -1) {
+      std::string s_copy(s.to_string());
+      re2::RE2::GlobalReplace(&s_copy, regex_replacement, replacement);
+      RETURN_NOT_OK(builder->Append(reinterpret_cast<const uint8_t*>(s_copy.data()),
+                                    s_copy.length()));
+      return Status::OK();
+    }
+    // Since RE2 does not have the concept of max_replacements, we have to do some work
+    // ourselves.
```

Comment:
Note that the `GlobalReplace` loop works a bit differently, it calls `Match` then `Rewrite`, avoiding the duplicate matching calls. Not sure it's worth optimizing this, though:
https://github.com/google/re2/blob/master/re2/re2.cc#L427

Reply:
Good idea, I prefer to keep it as it is, I left a comment in the code so this doesn't get lost.